ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/BatchingParallelIngestionPipe.scala

```diff
@@ -42,9 +40,36 @@ object BatchingParallelIngestionPipe {
       .mapAsync(ingestingParallelism)(ingester)
       // Stage 6: Preparing data sequentially for throttled mutations in database (tracking the ledger-end, corresponding sequential event ids and latest-at-the-time configurations)
       .conflate(tailer)
-      .throttle(tailingRateLimitPerSecond, FiniteDuration(1, "seconds"))
       // Stage 7: Updating ledger-end and related data in database (this stage completion demarcates the consistent point-in-time)
       .mapAsync(1)(ingestTail)
       .map(_ => ())
 
+  /** Forms dynamically-sized batches based on downstream backpressure.
+    *   - Under light load, this flow emits batches of size 1.
+    *   - Under heavy load, this flow emits batches of `maxBatchSize`.
+    */
+  private def batchN[IN](
+      maxBatchSize: Int,
+      maxBatchCount: Int,
+  ): Flow[IN, ArrayBuffer[IN], NotUsed] =
+    Flow[IN]
+      .batch[Vector[ArrayBuffer[IN]]](
+        (maxBatchSize * maxBatchCount).toLong,
+        in => Vector(newBatch(maxBatchSize, in)),
+      ) { case (batches, in) =>
+        val lastBatch = batches.last
+        if (lastBatch.size < maxBatchSize) {
+          lastBatch.addOne(in)
+          batches
+        } else
+          batches :+ newBatch(maxBatchSize, in)
+      }
+      .flatMapConcat(v => Source.fromIterator(() => v.iterator))
```

Comment:
would .mapConcat(identity) work here?
In some recent experiments I observed a mapConcat is faster/cheaper

Reply:
Didn't think of it. Will try